ddtrace/internal/_encoding.pyx

```diff
@@ -112,38 +123,160 @@ cdef inline int pack_text(msgpack_packer *pk, object text):
     raise TypeError("Unhandled text type: %r" % type(text))
 
 
-cdef class Packer(object):
-    """
-    MessagePack Packer
+cdef class BufferedEncoder(object):
+    content_type: str = None
+
+    cdef public int max_size
+    cdef public int max_item_size
+    cdef object _lock
+
+    def __cinit__(self, max_size, max_item_size):
+        self.max_size = max_size
+        self.max_item_size = max_item_size
+        self._lock = threading.Lock()
+
+    # ---- Abstract methods ----
+
+    def put(self, item):
+        raise NotImplementedError()
+
+    def encode(self):
+        raise NotImplementedError()
 
-    usage::
 
-        packer = Packer()
-        astream.write(packer.pack_trace(trace))
-        astream.write(packer.pack_traces(traces))
+cdef class MsgpackEncoderBase(BufferedEncoder):
+    content_type = "application/msgpack"
 
-    """
     cdef msgpack_packer pk
-    cdef const char *encoding
-    cdef const char *unicode_errors
+    cdef stdint.uint32_t _count
 
-    def __cinit__(self):
+    def __cinit__(self, max_size, max_item_size):
         cdef int buf_size = 1024*1024
         self.pk.buf = <char*> PyMem_Malloc(buf_size)
         if self.pk.buf == NULL:
             raise MemoryError("Unable to allocate internal buffer.")
+
+        self.max_size = max_size
         self.pk.buf_size = buf_size
-        self.pk.length = 0
+        self.max_item_size = max_item_size if max_item_size < max_size else max_size
+        self._lock = threading.Lock()
+        self._reset_buffer()
 
     def __dealloc__(self):
         PyMem_Free(self.pk.buf)
         self.pk.buf = NULL
 
-    cdef inline object _flush_buffer(self):
-        buf = PyBytes_FromStringAndSize(self.pk.buf, self.pk.length)
-        # Reset the buffer.
-        self.pk.length = 0
-        return buf
+    def __len__(self):  # TODO: Use a better name?
```

Comment:
what is wrong with this name?

Reply:
The buffered encoder keeps track of the number of traces that have been encoded as well as the current buffer utilisation. Here `__len__` is used for the number of traces but could be easily confused for the buffer utilisation, hence why I wasn't particularly happy with this choice.